File: app/src/main/java/com/moggot/findmycarlocation/map/GoogleMapFragment.java
package com.moggot.findmycarlocation.map;

import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatImageView;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.maps.android.PolyUtil;
import com.moggot.findmycarlocation.MainActivity;
import com.moggot.findmycarlocation.R;
import com.moggot.findmycarlocation.common.BaseFragment;
import com.moggot.findmycarlocation.common.ErrorStatus;

import java.util.Calendar;
import java.util.List;

import butterknife.BindView;

public class GoogleMapFragment extends BaseFragment<MapViewModel> implements OnMapReadyCallback {

    private static final int LOCATION_UPDATE_PERIOD = 10000;
    private static final int SUGGEST_PURCHASE_FREQUENCY = 3;

    private static final String TAG = "GoogleMapFragment";
    @BindView(R.id.tv_distance_value)
    TextView tvDistance;
    @BindView(R.id.tv_duration_value)
    TextView tvDuration;
    @BindView(R.id.map_tv_lat)
    TextView tvLat;
    @BindView(R.id.map_tv_lng)
    TextView tvLng;
    @BindView(R.id.map_btn_found)
    TextView btnFound;
    @BindView(R.id.map_iv_location)
    AppCompatImageView ivLocation;
    @BindView(R.id.map_status_dot)
    View viewDot;
    @BindView(R.id.map)
    MapView googleMapView;
    Handler handler = new Handler();
    @Nullable
    private GoogleMap map;
    private MapViewModel viewModel;
    private final Runnable runnableCode = new Runnable() {
        @Override
        public void run() {
            viewModel.getCurrentLocation();
            handler.postDelayed(this, LOCATION_UPDATE_PERIOD);
        }
    };

    public static GoogleMapFragment newInstance() {
        return new GoogleMapFragment();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        googleMapView.onCreate(savedInstanceState);
        googleMapView.getMapAsync(this);
        enableSearchMode(false);
        viewModel.getRouteData().observe(this, path -> {
            MainActivity activity = ((MainActivity) getActivity());
            if (activity == null) {
                return;
            }
            if (!activity.isPremiumPurchased()) {
                showInterstitial();
            }
            if (path == null) {
                return;
            }
            enableSearchMode(true);
            showDistance(path.getRoutes().get(0).getLegs().get(0).getDistance().getText());
            showDuration(path.getRoutes().get(0).getLegs().get(0).getDuration().getText());
            String pointsStr = path.getRoutes().get(0).getOverviewPolyline().getPoints();
            List<LatLng> points = PolyUtil.decode(pointsStr);
            showRoute(points);
        });
        viewModel.getErrorStatus().observe(this, errorStatus -> {
            if (errorStatus.getStatus() == ErrorStatus.BUILD_PATH_ERROR) {
                Snackbar.make(view, getString(R.string.no_path), Snackbar.LENGTH_INDEFINITE)
                        .setAction(getString(R.string.retry), action -> viewModel.retryCall())
                        .show();
            }
        });
        viewModel.getLocationData().observe(this, location -> {
            if (location == null) {
                return;
            }
            handler.post(runnableCode);
            tvLat.setText(String.valueOf(location.getLatitude()));
            tvLng.setText(String.valueOf(location.getLongitude()));
        });
        btnFound.setOnClickListener(v -> {
            viewModel.foundCar();
            if (getActivity() != null) {
                enableSearchMode(false);
                Toast.makeText(getContext(), getString(R.string.car_is_found), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState, MapViewModel viewModel) {
        this.viewModel = viewModel;
    }

    @Override
    protected Class<MapViewModel> getViewModel() {
        return MapViewModel.class;
    }

    @Override
    public String getFragmentTag() {
        return TAG;
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_map;
    }

    private void enableSearchMode(boolean enable) {
        btnFound.setVisibility(enable ? View.VISIBLE : View.INVISIBLE);
        btnFound.setEnabled(enable);
        viewDot.setBackgroundResource(enable ? R.drawable.status_green_dot : R.drawable.status_red_dot);
        tvDistance.setVisibility(enable ? View.VISIBLE : View.GONE);
        tvDuration.setVisibility(enable ? View.VISIBLE : View.GONE);
        if (enable) {
            Animation pulse = AnimationUtils.loadAnimation(getContext(), R.anim.pulse_scale);
            ivLocation.startAnimation(pulse);
            decoratePoint(viewModel.drawCircle());
        } else {
            ivLocation.clearAnimation();
            if (map != null) {
                map.clear();
            }
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        googleMapView.onStart();
    }

    @Override
    public void onResume() {
        super.onResume();
        googleMapView.onResume();
    }

    @Override
    public void onPause() {
        googleMapView.onPause();
        super.onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        googleMapView.onStop();
    }

    @Override
    public void onDestroyView() {
        googleMapView.onDestroy();
        ivLocation.clearAnimation();
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        viewModel.getRouteData().removeObservers(this);
        handler.removeCallbacks(runnableCode);
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        googleMapView.onLowMemory();
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        map = googleMap;
        if (ContextCompat.checkSelfPermission(getContext(),
                android.Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        if (map != null) {
            map.setMyLocationEnabled(true);
            map.getUiSettings().setZoomControlsEnabled(true);
            map.getUiSettings().setCompassEnabled(true);
            viewModel.buildRoute();
        }
    }

    private void decoratePoint(LatLng point) {
        drawCircle(point);
        addMarker(point);
    }

    private void drawCircle(LatLng point) {
        CircleOptions circleOptions = new CircleOptions();
        circleOptions.center(point);
        circleOptions.radius(10);
        circleOptions.strokeColor(Color.BLACK);
        circleOptions.fillColor(0x30ff0000);
        circleOptions.strokeWidth(2);
        if (map != null) {
            map.addCircle(circleOptions);
        }
    }

    private void addMarker(LatLng carPosition) {
        MarkerOptions endMarkerOptions = new MarkerOptions()
                .position(carPosition)
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.car));
        if (map != null) {
            map.addMarker(endMarkerOptions);
        }
    }

    private void showInterstitial() {
        InterstitialAd interstitialAd = new InterstitialAd(getContext());
        interstitialAd.setAdUnitId(getString(R.string.banner_ad_unit_id_map_interstitial));
        AdRequest adRequest = new AdRequest.Builder().build();
        interstitialAd.loadAd(adRequest);
        interstitialAd.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                if (interstitialAd.isLoaded()) {
                    interstitialAd.show();
                }
            }

            @Override
            public void onAdOpened() {
                //do nothing
            }

            @Override
            public void onAdFailedToLoad(int errorCode) {
                //do nothing
            }
        });
    }

    private void showRoute(List<LatLng> points) {
        if (map == null) {
            return;
        }
        PolylineOptions line = new PolylineOptions();
        line.width(4f).color(R.color.line);
        LatLngBounds.Builder latLngBuilder = new LatLngBounds.Builder();
        for (int i = 0; i < points.size(); i++) {
            line.add(points.get(i));
            latLngBuilder.include(points.get(i));
        }
        map.addPolyline(line);
        int size = getResources().getDisplayMetrics().widthPixels;
        LatLngBounds latLngBounds = latLngBuilder.build();
        CameraUpdate track = CameraUpdateFactory.newLatLngBounds(latLngBounds, size, size, 25);
        map.animateCamera(track);
    }

    private void showDistance(String distance) {
        tvDistance.setText(getString(R.string.distance, distance));
    }

    private void showDuration(String duration) {
        tvDuration.setText(getString(R.string.duration, duration));
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/map/GoogleMapFragment.kt
package com.moggot.findmycarlocation.map

import android.content.pm.PackageManager
import android.graphics.Color
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.InterstitialAd
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.*
import com.google.android.material.snackbar.Snackbar
import com.google.maps.android.PolyUtil
import com.moggot.findmycarlocation.MainActivity
import com.moggot.findmycarlocation.R
import com.moggot.findmycarlocation.base.BaseFragment
import com.moggot.findmycarlocation.data.model.route.Path
import com.moggot.findmycarlocation.di.ComponentHolder
import com.moggot.findmycarlocation.di.component.AppComponent
import com.moggot.findmycarlocation.di.component.MainComponent
import javax.inject.Inject

class GoogleMapFragment : BaseFragment(), OnMapReadyCallback, GoogleMapView {

    lateinit var tvDistance: TextView
    lateinit var tvDuration: TextView
    lateinit var tvLat: TextView
    lateinit var tvLng: TextView
    lateinit var btnFound: TextView
    lateinit var ivLocation: AppCompatImageView
    lateinit var viewDot: View
    lateinit var googleMapView: MapView
    private lateinit var adView: AdView
    private lateinit var adRequest: AdRequest

    private val handler = Handler()
    private var map: GoogleMap? = null
    private val runnableCode = object : Runnable {
        override fun run() {
            handler.postDelayed(this, LOCATION_UPDATE_PERIOD.toLong())
        }
    }
    @Inject
    lateinit var presenter: MapPresenter

    override val fragmentTag: String
        get() = TAG

    override val componentName: String
        get() = AppComponent::class.java.name

    override val isComponentDestroyable: Boolean
        get() = false

    override val layoutRes: Int
        get() = R.layout.fragment_map

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adRequest = AdRequest.Builder().build()
        ComponentHolder.provideComponent(MainComponent::class.java.name) {
            MainComponent.Initializer.init()
        }.inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter.attachView(this)
        tvDistance = view.findViewById(R.id.tv_distance_value)
        tvDuration = view.findViewById(R.id.tv_duration_value)
        tvLat = view.findViewById(R.id.map_tv_lat)
        tvLng = view.findViewById(R.id.map_tv_lng)
        btnFound = view.findViewById(R.id.map_btn_found)
        ivLocation = view.findViewById(R.id.map_iv_location)
        viewDot = view.findViewById(R.id.map_status_dot)
        googleMapView = view.findViewById(R.id.map)
        adView = view.findViewById(R.id.ad_map)
        val activity = activity as MainActivity
        if (!activity.isPremiumPurchased) {
            adView.loadAd(adRequest)
        }

        googleMapView.onCreate(savedInstanceState)
        googleMapView.getMapAsync(this)
        enableSearchMode(false)
        presenter.getCurrentLocation()
        btnFound.setOnClickListener {
            presenter.foundCar()
            enableSearchMode(false)
        }
    }

    override fun drawRoute(path: Path?) {
        if (!(activity as MainActivity).isPremiumPurchased) {
            showInterstitial()
        }
        if (path == null) {
            return
        }
        enableSearchMode(true)
        showDistance(path.routes[0].legs[0].distance.text)
        showDuration(path.routes[0].legs[0].duration.text)
        val pointsStr = path.routes[0].overviewPolyline.points
        val points = PolyUtil.decode(pointsStr)
        showRoute(points)
    }

    private fun enableSearchMode(enable: Boolean) {
        btnFound.visibility = if (enable) View.VISIBLE else View.INVISIBLE
        btnFound.isEnabled = enable
        viewDot.setBackgroundResource(if (enable) R.drawable.status_green_dot else R.drawable.status_red_dot)
        tvDistance.visibility = if (enable) View.VISIBLE else View.GONE
        tvDuration.visibility = if (enable) View.VISIBLE else View.GONE
        if (enable) {
            val pulse = AnimationUtils.loadAnimation(context, R.anim.pulse_scale)
            ivLocation.startAnimation(pulse)
            decoratePoint(presenter.drawCircle())
        } else {
            ivLocation.clearAnimation()
            map?.clear()
        }
    }

    override fun onStart() {
        super.onStart()
        googleMapView.onStart()
    }

    override fun onResume() {
        super.onResume()
        googleMapView.onResume()
        adView.resume()
    }

    override fun onPause() {
        googleMapView.onPause()
        adView.pause()
        super.onPause()
    }

    override fun onStop() {
        super.onStop()
        googleMapView.onStop()
    }

    override fun onDestroyView() {
        googleMapView.onDestroy()
        ivLocation.clearAnimation()
        presenter.detachView()
        adView.destroy()
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
        handler.removeCallbacks(runnableCode)
    }

    override fun onLowMemory() {
        super.onLowMemory()
        googleMapView.onLowMemory()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap
        if (ContextCompat.checkSelfPermission(requireContext(),
                        android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return
        }
        map?.isMyLocationEnabled = true
        map?.uiSettings?.isZoomControlsEnabled = true
        map?.uiSettings?.isCompassEnabled = true
        presenter.buildRoute()
    }

    private fun decoratePoint(point: LatLng) {
        drawCircle(point)
        addMarker(point)
    }

    private fun drawCircle(point: LatLng) {
        val circleOptions = CircleOptions()
        circleOptions.center(point)
        circleOptions.radius(10.0)
        circleOptions.strokeColor(Color.BLACK)
        circleOptions.fillColor(0x30ff0000)
        circleOptions.strokeWidth(2f)
        map?.addCircle(circleOptions)
    }

    private fun addMarker(carPosition: LatLng) {
        val endMarkerOptions = MarkerOptions()
                .position(carPosition)
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.car))
        map?.addMarker(endMarkerOptions)
    }

    private fun showInterstitial() {
        val interstitialAd = InterstitialAd(context)
        interstitialAd.adUnitId = getString(R.string.banner_ad_unit_id_map_interstitial)
        val adRequest = AdRequest.Builder().build()
        interstitialAd.loadAd(adRequest)
        interstitialAd.adListener = object : AdListener() {
            override fun onAdLoaded() {
                if (interstitialAd.isLoaded) {
                    interstitialAd.show()
                }
            }

            override fun onAdOpened() {
                //do nothing
            }

            override fun onAdFailedToLoad(errorCode: Int) {
                //do nothing
            }
        }
    }

    private fun showRoute(points: List<LatLng>) {
        val line = PolylineOptions()
        line.width(4f).color(R.color.line)
        val latLngBuilder = LatLngBounds.Builder()
        for (i in points.indices) {
            line.add(points[i])
            latLngBuilder.include(points[i])
        }
        map?.addPolyline(line)
        val size = resources.displayMetrics.widthPixels
        val latLngBounds = latLngBuilder.build()
        val track = CameraUpdateFactory.newLatLngBounds(latLngBounds, size, size, 25)
        map?.animateCamera(track)
    }

    private fun showDistance(distance: String) {
        tvDistance.text = getString(R.string.distance, distance)
    }

    private fun showDuration(duration: String) {
        tvDuration.text = getString(R.string.duration, duration)
    }

    override fun updateLocation(location: Location?) {
        handler.post(runnableCode);
        tvLat.text = location?.latitude.toString()
        tvLng.text = location?.longitude.toString()
    }

    override fun onError(throwable: Throwable) {
        Snackbar.make(googleMapView, getString(R.string.no_path), Snackbar.LENGTH_INDEFINITE)
                .setAction(getString(R.string.retry)) { presenter.retryCall() }
                .show()
    }

    override fun onCarFound() {
        Toast.makeText(context, getString(R.string.car_is_found), Toast.LENGTH_SHORT).show()
    }

    companion object {

        private const val LOCATION_UPDATE_PERIOD = 10000
        private const val TAG = "GoogleMapFragment"

        fun newInstance(): GoogleMapFragment {
            return GoogleMapFragment()
        }
    }
}
